### ISSUE TITLE
ISSUE: Keycodes stuck on Disconnecting when any advanced options are set

### ISSUE THREAD

**Issue by thayerfox** [FutureTense/keymaster#121]
I am having an issue after the recent updates. I am running 2021.3.1 and 0.0.27  (issue was originally noticed on 0.0.26), with the old ZWave integration on a Kwikset lock. If I set any of the advanced options for pin state the PIN Status switches to Disconnecting and PIN synchronized with lock is always Off.  (If I disable whatever advanced option triggered the disconnecting, it connects, reenables and syncs fine.)

For example:

- Enable Enforce PIN limit and set Unlock events to 1. PIN Status changes to Connected, PIN synched On.
- Set Unlock events to 0. PIN Status changes to Disconnecting, PIN synched Off. This never changes to Disconnected and synched.
- Set Unlock events to 1. PIN Status changes to Connected, PIN synched On.

The same behavior exists for any of the other options, i.e., if I set the pin to not be currently valid for the current date or time, it gets stuck at disconnecting. Reset Code Slot works fine though.

This was working a few days ago, before the updates. I have tried uninstalling, removing the repo, deleting any remaining files or entities and reinstalling, but I get the same behavior.

**Comment by raman325** [comment]
hi @thayerfox do you recall what version worked? It would help narrow down what could be causing this

**Comment by thayerfox** [comment]
I last tested the advanced trigger code disconnect on March 1st @4pm pst, using whatever the latest release was at the time.  Looking at the release log, 0.0.22 was released Mar 1st @10am pst.

Are there any logs or something I can dig through to provide more info?

**Comment by raman325** [comment]
if you set the log level to debug, you will see more information about what's happening

**Comment by raman325** [comment]
and to clarify, are you saying that it **did** work on March 1st during your test and that you were updated to the latest?

Have you also regenerated your package files since upgrading?

**Comment by thayerfox** [comment]
Correct, it worked on March 1st.  I just installed it that day, so it would have been the latest version at that time.

Package files have been regenerated, as well as a full removal and reinstallation of the add-on.

**Comment by thayerfox** [comment]
Setting valid code in position 1 via "Enabled" toggle.  (note, whatever is in postion 3 is not valid, I tried factory resetting the lock to clear it (which worked for other key positions) but not #3.  The code is not valid on the lock.  Within 30 sec, the pin status changes to connected and pin sync with lock is true.)
> 2021-03-15 16:08:18 DEBUG (MainThread) [custom_components.keymaster] Add Code service: <ServiceCall keymaster.add_code (c:f367558f8f3013234d5f71ddb326683b): entity_id=lock.front_door_lock_locked, code_slot=1, usercode=1234>
2021-03-15 16:08:18 DEBUG (MainThread) [custom_components.keymaster.services] Attempting to call set_usercode...
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 255 value: 30
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 254 value: False
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 0 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 1 value: 3988
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 2 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 3 value: 6747
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 4 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 5 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 6 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 7 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 8 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 9 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 10 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 11 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 12 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 13 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 14 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 15 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 16 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 17 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 18 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 19 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 20 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 21 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 22 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 23 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 24 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 25 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 26 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 27 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 28 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 29 value: 
2021-03-15 16:08:19 DEBUG (SyncWorker_4) [custom_components.keymaster] DEBUG: Code slot 30 value: 
2021-03-15 16:08:19 DEBUG (MainThread) [custom_components.keymaster] Finished fetching keymaster data in 0.032 seconds
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 255 value: 30
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 254 value: False
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 0 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 1 value: 1234
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 2 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 3 value: 6747
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 4 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 5 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 6 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 7 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 8 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 9 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 10 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 11 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 12 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 13 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 14 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 15 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 16 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 17 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 18 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 19 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 20 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 21 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 22 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 23 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 24 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 25 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 26 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 27 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 28 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 29 value: 
2021-03-15 16:08:24 DEBUG (SyncWorker_1) [custom_components.keymaster] DEBUG: Code slot 30 value: 
2021-03-15 16:08:24 DEBUG (MainThread) [custom_components.keymaster] Finished fetching keymaster data in 0.018 seconds

Clearing the code, via toggling the Enabled toggle (shows disconnecting, then disconnected and synch with lock true):

> 2021-03-15 16:08:53 DEBUG (MainThread) [custom_components.keymaster] Clear Code service: <ServiceCall keymaster.clear_code (c:f6c83018e2fde761bba099d6f128d14b): entity_id=lock.front_door_lock_locked, code_slot=1>
2021-03-15 16:08:53 DEBUG (MainThread) [custom_components.keymaster.services] Attempting to call clear_usercode...
2021-03-15 16:08:53 DEBUG (MainThread) [custom_components.keymaster.services] Setting code slot value to random PIN as workaround in case clearing code doesn't work
2021-03-15 16:08:53 ERROR (MainThread) [homeassistant.components.google_assistant.http] Request for https://homegraph.googleapis.com/v1/devices:reportStateAndNotification failed: 404
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 255 value: 30
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 254 value: False
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 0 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 1 value: 1234
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 2 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 3 value: 6747
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 4 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 5 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 6 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 7 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 8 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 9 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 10 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 11 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 12 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 13 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 14 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 15 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 16 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 17 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 18 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 19 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 20 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 21 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 22 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 23 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 24 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 25 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 26 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 27 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 28 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 29 value: 
2021-03-15 16:08:54 DEBUG (SyncWorker_0) [custom_components.keymaster] DEBUG: Code slot 30 value: 
2021-03-15 16:08:54 DEBUG (MainThread) [custom_components.keymaster] Finished fetching keymaster data in 0.020 seconds
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 255 value: 30
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 254 value: False
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 0 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 1 value: 1245
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 2 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 3 value: 6747
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 4 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 5 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 6 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 7 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 8 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 9 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 10 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 11 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 12 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 13 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 14 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 15 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 16 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 17 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 18 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 19 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 20 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 21 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 22 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 23 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 24 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 25 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 26 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 27 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 28 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 29 value: 
2021-03-15 16:08:59 DEBUG (SyncWorker_11) [custom_components.keymaster] DEBUG: Code slot 30 value: 
2021-03-15 16:08:59 DEBUG (MainThread) [custom_components.keymaster] Finished fetching keymaster data in 0.017 seconds

Renabling and then toggling the current day under advanced options (shows disconnecting, never changes status on UI, pin sync with lock is false.  I checked, old code, 1234 no longer works, new random code from log now does):

> 2021-03-15 16:18:04 DEBUG (MainThread) [custom_components.keymaster] Clear Code service: <ServiceCall keymaster.clear_code (c:a4ff748fdaa5681053d08a226f32542e): entity_id=lock.front_door_lock_locked, code_slot=1>
2021-03-15 16:18:04 DEBUG (MainThread) [custom_components.keymaster.services] Attempting to call clear_usercode...
2021-03-15 16:18:04 DEBUG (MainThread) [custom_components.keymaster.services] Setting code slot value to random PIN as workaround in case clearing code doesn't work
2021-03-15 16:18:04 ERROR (MainThread) [homeassistant.components.google_assistant.http] Request for https://homegraph.googleapis.com/v1/devices:reportStateAndNotification failed: 404
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 255 value: 30
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 254 value: False
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 0 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 1 value: 7187
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 2 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Utilizing Zwave clear_usercode work around code.
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 3 value: 6747
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 4 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 5 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 6 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 7 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 8 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 9 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 10 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 11 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 12 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 13 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 14 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 15 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 16 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 17 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 18 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 19 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 20 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 21 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 22 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 23 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 24 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 25 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 26 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 27 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 28 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 29 value: 
2021-03-15 16:18:09 DEBUG (SyncWorker_2) [custom_components.keymaster] DEBUG: Code slot 30 value: 
2021-03-15 16:18:09 DEBUG (MainThread) [custom_components.keymaster] Finished fetching keymaster data in 0.016 seconds

Toggling the date back to true results in the UI changing from disconnecting to connecting.  The logs show the PIN never changes back though.  If I toggle Enabled to be false, and then toggle it back true, it connects and syncs back.

So, in summary:

-It appears that disconnecting a pin via an advanced option does remove the pin from the lock.
-Reenabling a pin which is still in the disconnecting state changes it to connecting, but does not enable it in the lock.

**Comment by firstof9** [comment]
Since you are using the old `zwave` integration, your code slots will get a randomly generated code when you disable them. This is due to a limitation in the old `zwave` integration, there is nothing we can do other than to set randomly generated codes that you would normally not be able to see, and this is  what you are currently seeing in play here.

It is recommended to upgrade to `zwave_js` for proper lock code management.

**Comment by thayerfox** [comment]
I understand the random key codes.  The issue is the status does not change from disconnecting if an advanced option disabled the code, and then subsequently doesn't re-enable the code without fully disabling the key slot.  (So if you have a code only valid on Mondays, it would show disconnecting, and then never re-enable on Tuesday)

What conditions are required for the integration to show disconnected and sync with lock true?  What is different about the advanced option driven disconnect versus the main code slot disconnect?

**Comment by firstof9** [comment]
> What conditions are required for the integration to show disconnected and sync with lock true?

```yaml
    connected_frontdoor_1:
      friendly_name: "PIN Status"
      value_template: >-
        {% set pin_active = is_state('binary_sensor.active_frontdoor_1', 'on')  %}      
        {% set synched = is_state('binary_sensor.pin_synched_frontdoor_1', 'on')  %}      
        {% if pin_active %}
          {% if synched %}
            Connected
          {% else %}
            Connecting
          {% endif %}
        {% else %}
          {% if synched %}
            Disconnected
          {% else %}
            Disconnecting
          {% endif %}
        {% endif %}
```



> What is different about the advanced option driven disconnect versus the main code slot disconnect?

Are you referring to "Reset Code Slot" ?

**Comment by thayerfox** [comment]
I am referring to input_boolean.enabled_frontdoor_1 -> using this toggle results in correct disconnecting>disconnected or connecting>connected status of sensor.connected_frontdoor_1 and binary_sensor.pin_synched_frontdoor_1 going false>true as expected.

If for example, I have a code enabled, and then toggle Enforce PIN limit with Unlock events 0, binary_sensor.pin_synched_frontdoor_1 remains false.  binary_sensor.active_frontdoor_1 toggles to false. 

If I then set Unlock events to 1, binary_sensor.pin_synched_frontdoor_1 remains false.  binary_sensor.active_frontdoor_1 toggles to true.

It will remain in this state, until I toggle input_boolean.enabled_frontdoor_1.  Toggling it off results in binary_sensor.pin_synched_frontdoor_1 going back to true 2 seconds later.  ( I can then toggle enabled_frontdoor_1 back to true, and have the code reenabled.)

It appears that binary_sensor.pin_synched_frontdoor_1 does not change to true after an advanced option attempts to disable a code.

**Comment by firstof9** [comment]
> It appears that binary_sensor.pin_synched_frontdoor_1 does not change to true after an advanced option attempts to disable a code.

I see.

**Comment by thayerfox** [comment]
Dug a little deeper.

>pin_synched_LOCKNAME_TEMPLATENUM:
      friendly_name: 'PIN synchronized with lock'
      value_template: >
        {% set lockpin = states('sensor.LOCKNAME_code_slot_TEMPLATENUM').strip()  %}
        {% set localpin = states('input_text.LOCKNAME_pin_TEMPLATENUM').strip()  %}
        {% set pin_active = is_state('binary_sensor.active_LOCKNAME_TEMPLATENUM', 'on')  %}
        {% if lockpin == "0000" %}
        {%   set lockpin = "" %}
        {% endif %}
        {% if pin_active %}
          {{ localpin == lockpin }}
        {% else %}
          {{ lockpin == "" }}
        {% endif %}

So, if I understand correctly:
sensor.frontdoor_code_slot_1 is the zwave reported codeslot. 
input_text.frontdoor_pin_1 is the desired pin in UI.

On a disable via input_boolean.enabled_frontdoor_1:

> sensor.frontdoor_code_slot_1 state = <blank>
input_text.frontdoor_pin_1  state = 1234 (the pin in the UI)
binary_sensor.active_frontdoor_1 state = false

On a disconnect via advanced, entry count 0 for example:

> sensor.frontdoor_code_slot_1 state = 2250 (random from zwave int)
input_text.frontdoor_pin_1  state = 1234 (the pin in the UI)
binary_sensor.active_frontdoor_1 state = true

so, in this case, the pin_active is true, and thus the localpin=lockpin comparison fails.  That would in turn set pin_synced_frontdoor_1 to false, no?

**Comment by raman325** [comment]
yes, when the states of the sensor and input_text are 2250 and 1234 respectively, pin_synched should be false.

I see a potential issue with the code but I can't see how it would have worked before. Regardless, can you try pulling down this branch and installing it manually instead (https://github.com/raman325/keymaster/tree/zwave_issue)? You won't need to regenerate package files, it's strictly a code change. Make sure that when you clone this repo down, that you checkout the `zwave_issue` branch before you copy it into your HA installation

**Comment by thayerfox** [comment]
Raman, that branch solves the issue for me.  Thanks!

As an aside, it was mentioned the random code instead of clearing was a workaround for the deprecated zwave integration.  According to my lock documentation, codes can only be cleared by overwriting with a different code.  How would that work with zwavejs?

> Deleting User Codes
In order to delete a user code, you must override the code by adding a different user code in the same position. For example, if
you want to delete the third code, add a different user code in position three.
If you cannot remember the user code position, you may wish to perform a factory reset to delete all codes associated with
the lock. See section “G.”
http://www.weiserlock.com/PDFs/instructions/Electronics/SmartCode_5_Installation_and_User_Guide_English.pdf (Section D-d)

**Comment by raman325** [comment]
sweet! Will push out a fix shortly.

We can probably add similar logic in the `zwave_js` integration. I will save that for a future release
